magicgui/function_gui.py

```diff
@@ -0,0 +1,286 @@
+    def __repr__(self) -> str:
+        """Return string representation of instance."""
+        fname = f"{self._function.__module__}.{self._function.__name__}"
+        return f"<FunctionGui {fname}{self.to_signature()}>"
+
+    @property
+    def result_name(self) -> str:
+        """Return a name that can be used for the result of this magicfunction."""
+        return self._result_name or (self._function.__name__ + " result")
+
+    @result_name.setter
+    def result_name(self, value: str):
+        """Set the result name of this MagicGui widget."""
+        self._result_name = value
+
+    def Gui(self, show=False):
+        """Create a widget instance [DEPRECATED]."""
+        warnings.warn(
+            "\n\nCreating a widget instance with `my_function.Gui()` is deprecated,\n"
+            "the magicgui decorator now returns a widget instance directly, so you\n"
+            "should simply use the function itself as a magicgui widget, or call\n"
+            "`my_function.show(run=True)` to run the application.\n"
+            "In a future version, the `Gui` attribute will be removed.\n",
+            FutureWarning,
+        )
+        if show:
+            self.show()
+        return self
+
+
+# ==================   magicgui decorator   ===================================
+
+F = TypeVar("F", bound=Callable[..., Any])
+
+
+@overload
+def magicgui(function: F, **k) -> FunctionGui:  # noqa: D103
+    ...
+
+
+@overload
+def magicgui(function=None, **k) -> Callable[[F], FunctionGui]:  # noqa: D103
+    ...
+
+
+def magicgui(
+    function: Optional[F] = None,
+    *,
+    orientation: str = "horizontal",
+    labels: bool = True,
+    call_button: Union[bool, str] = False,
+    auto_call: bool = False,
+    result_widget: bool = False,
+    app: AppRef = None,
+    **param_options: dict,
+):
+    """Create a FunctionGui class for ``function`` and add it as an attribute ``Gui``.
+
+    Parameters
+    ----------
+    function : Callable, optional
+        The function to decorate.  Optional to allow bare decorator with optional
+        arguments. by default None
+    orientation : str, optional
+        The type of layout to use. Must be one of {'horizontal', 'vertical',
+        'form', 'grid'}, by default "horizontal"
+    labels : bool
+        Whether labels are shown in the widget. by default True
+    call_button : bool or str, optional
+        If True, create an additional button that calls the original function when
+        clicked.  If a ``str``, set the button text. by default False
+    auto_call : bool, optional
+        If True, changing any parameter in either the GUI or the widget attributes
+        will call the original function with the current settings. by default False
+
+    **param_options : dict of dict
+        Any additional keyword arguments will be used as parameter-specific options.
+        Keywords MUST match the name of one of the arguments in the function
+        signature, and the value MUST be a dict.
```

Comment:
Am I correct in thinking that the values dicts must line up for with the acceptable widget types. Ultimately we'll want to link out here to the available widgets etc. (I know it's early to be thinking about that, and that can come later, just making sure I'm parsing everything critical right).

Reply:
Yeah, good intuition.  You're exactly right.  Not all key/values will work for all widget types.  I [started](https://github.com/napari/magicgui/blob/8965c9402ddb74b323633dd01de30453e5072ff2/magicgui/types.py#L46) on creating some typing support for that, widget-specific options still needs work/clarification.